### PR TITLE
Document the commands JSON files

### PR DESCRIPTION
**Pull request by zuiderkwast**
These JSON files were originally not intended to be used directly, since they
contain internals and some fiels like "acl_categories" that are not the final
ACL categories. (Valkey will apply some implicit rules to compute the final ACL
categories.) However, people see JSON files and use them directly anyway.

So it's better to document them.

In a later PR, we can get rid of all implicit ACL categories and instead populate them explicitly in the JSON files. Then, we'll add a validation (e.g. in generate-command-code.py) that the implied categories are set.

### CODE REVIEW

File: src/commands/README.md [hunk 1]
```diff
@@ -1,13 +1,216 @@
...
+
+Structure
+---------
+
+Each JSON file contains an object with a single key. The key is the command name
+in uppercase, e.g. "HSCAN" (hscan.json). The value is a JSON object with the
+following keys. To be safe, assume all of them are optional.
+
+* `"summary"`: a string with a short description of the command. One sentence.
+* `"complexity`: a string like `"O(1)"` or longer, like `"O(1) for every call. O(N) for a complete iteration, including enough command calls for the cursor to return back to 0. N is the number of elements inside the collection."`.
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
* `"complexity"`: a string like `"O(1)"` or longer, like `"O(1) for every call. O(N) for a complete iteration, including enough command calls for the cursor to return back to 0. N is the number of elements inside the collection."`.
```

File: src/commands/README.md [hunk 2]
```diff
@@ -1,13 +1,216 @@
...
+* `"since"`: a string with a version number, like "7.0.0". It's the version
+  (Redis OSS or Valkey) where the command was introduced.
+* `"arity"`: The number of arguments, including the command name itself. A
+  negative number means "at least", e.g. -3 means at least 3.
+* `"container"`: Only present for subcommands. See below.
+* `"history"`: An array of changes, each change represented by a 2-element array
+  on the form `[VERSION, DESCRIPTION]`.
+* `"function"`: The name of the C function in Valkey's source code implementing
+  the command. (Don't use it for anything else.)
+* `command_flags`: An array of flags represented as strings. Command flags:
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
* `"command_flags"`: An array of flags represented as strings. Command flags:
```

File: src/commands/README.md [hunk 3]
```diff
@@ -1,13 +1,216 @@
...
+    * `"block"`: A group of arguments. The elements are in the key `"arguments"`.
+    * `"double"`: A number, not necessarily an integer.
+    * `"integer"`: An integer.
+    * `"key"`: A string representing a key in the database.
+    * `"oneof"`: One of a list of alternatives. The alternatives are in the key `"arguments"`.
+    * `"pattern"`: A string representing a glob-style pattern.
+    * `"pure-token"`: A fixed string. The string is in the key `"token"`.
+    * `"string"`: A string.
+    * `"unix-time"`: An integer representing a unix time in either seconds or milliseconds.
+  * `"arguments"`: A list with the same structure as its parent. Present if type is "block" or "oneof"
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
  * `"arguments"`: A list with the same structure as its parent. Present if type is "block" or "oneof".
```

File: src/commands/README.md [hunk 4]
```diff
@@ -1,13 +1,216 @@
...
+  * `"display"`: ("entries-read", "key" or "pattern")
+  * `"key_spec_index"`: An index into the `key_specs` array. Only if `type` is `key`.
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
  * `"key_spec_index"`: An index into the `"key_specs"` array. Only if `"type"` is `"key"`.
```

File: src/commands/README.md [hunk 5]
```diff
@@ -1,13 +1,216 @@
...
+  * `"multiple":`  true if the argument can be repeated multiple times. Omitted means false.
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
  * `"multiple":` true if the argument can be repeated multiple times. Omitted means false.
```

File: src/commands/README.md [hunk 6]
```diff
@@ -1,13 +1,216 @@
...
+  * `"optional":` True if the argument is optional. Omitted means false.
+  * `"since"`: Version (string) when the argument was introduced.
+  * `"token"`: A string indicating a fixed string value. This is always present
+    if type is "pure-token". If type is anything else, then `"token"` indicates
+    the argument is preceded by an extra (fixed string) argument.
+
+Implicit ACL categories
+-----------------------
+
+The ACL categories specified as `acl_categories` are not the ones actually used.
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
The ACL categories specified as `"acl_categories"` are not the ones actually used.
```

File: src/commands/README.md [hunk 7]
```diff
@@ -1,13 +1,216 @@
...
+  * `"NOT_KEY"`
+  * `"OW"`
+  * `"RM"`
+  * `"RO"`
+  * `"RW"`
+  * `"UPDATE"`
+  * `"VARIABLE_FLAGS"`
+* `"begin_search"`: How to find the first key used by this key spec. It's an
+  object with only one key. The key determines the method for finding the first
+  key. Here are the possible forms of the `begin_search` object:
```
Comment (enjoy-binbin) [inline src/commands/README.md]:
```suggestion
  key. Here are the possible forms of the `"begin_search"` object:
```

**Review by enjoy-binbin — APPROVED** [review]

File: src/commands/README.md (hunk 1 above)
Comment (zuiderkwast) [reply src/commands/README.md]: Thank you, master to typos!

File: src/commands/README.md [hunk 8]
```diff
@@ -1,13 +1,216 @@
...
 
-Each JSON contains all the information about the command itself, but these JSON files are not to be used directly!
-Any third party who needs access to command information must get it from `COMMAND INFO` and `COMMAND DOCS`.
-The output can be extracted in a JSON format by using `redis-cli --json`, in the same manner as in `utils/generate-commands-json.py`.
+Each JSON file contains all the information about the command itself. It is the
+"single source of truth" (SSOT) for the command's metadata.
+
+These JSON files were originally not intended to be used directly, since they
+contain internals and some fields like "acl_categories" that are not the final
+ACL categories. (Valkey will apply some implicit rules to compute the final ACL
```
Comment (stockholmux) [inline src/commands/README.md:11]: Maybe we should make an issue about the implicit rules and reference the issue here so we have a paper trail?

File: src/commands/README.md [hunk 9]
```diff
@@ -1,13 +1,216 @@
...
+  * `"sorted_set"`
+  * `"stream"`
+  * `"string"`
+  * `"transactions"`
+* `"since"`: a string with a version number, like "7.0.0". It's the version
+  (Redis OSS or Valkey) where the command was introduced.
+* `"arity"`: The number of arguments, including the command name itself. A
+  negative number means "at least", e.g. -3 means at least 3.
+* `"container"`: Only present for subcommands. See below.
+* `"history"`: An array of changes, each change represented by a 2-element array
```
Comment (stockholmux) [inline src/commands/README.md:51]: Might be worth noting that should be omitted if empty. Previously Redis inconsistently used a bunch of `[]` entriee.

**Review by stockholmux — CHANGES_REQUESTED** [review]
Great to see this. A couple of small issues.

File: src/commands/README.md (hunk 8 above)
Comment (zuiderkwast) [reply src/commands/README.md:11]: Sure, will do

File: src/commands/README.md (hunk 9 above)
Comment (zuiderkwast) [reply src/commands/README.md:51]: Sure. That applies to all arrays here, probably. We don't have any validation for it though. We can add that in one of the scripts too.

**Pull request merged**